vnext/Desktop/Modules/BlobModule.cpp

```diff
@@ -0,0 +1,239 @@
+
+        if (auto instance = getInstance().lock())
+        {
+          auto buffer = CryptographicBuffer::CreateFromByteArray(data);
+          auto winrtString = CryptographicBuffer::EncodeToBase64String(std::move(buffer));
+          auto base64String = Common::Unicode::Utf16ToUtf8(std::move(winrtString));
+
+          auto sendArgs = dynamic::array(std::move(base64String), socketID);
+          instance->callJSFunction("WebSocketModule", "sendBinary", std::move(sendArgs));
+        }
+      }
+    ),
+
+    Method(
+      "createFromParts",
+      [this](dynamic args)
+      {
+        auto parts = jsArgAsArray(args, 0);   // Array<Object>
+        auto blobId = jsArgAsString(args, 1);
+        vector<uint8_t> buffer{};
+
+        for(auto& part : parts)
+        {
+          auto type = part["type"];
+          if (type == "blob")
+          {
+            auto blob = part["data"];
+            auto blobId = blob["blobId"].asString();
+
+            auto bufferPart = m_contentHandler->ResolveMessage(blob["blobId"].asString(), blob["offset"].asInt(), blob["size"].asInt());
+            buffer.reserve(buffer.size() + bufferPart.size());
+            buffer.insert(buffer.end(), bufferPart.begin(), bufferPart.end());
+          }
+          else if (type == "string")
+          {
+            auto data = part["data"].asString();
+            auto bufferPart = vector<uint8_t>(data.begin(), data.end());
+
+            buffer.reserve(buffer.size() + bufferPart.size());
+            buffer.insert(buffer.end(), bufferPart.begin(), bufferPart.end());
+          }
+          else// if (auto instance = getInstance().lock())
+          {
+            //TODO: Error?
+            //instance->callJSFunction("RCTDeviceEventEmitter", "emit", dynamic::array("eventName", dynamic{}));
+            return;
+          }
+
+          m_contentHandler->StoreMessage(std::move(buffer), std::move(blobId));
+        }
+      }
+    ),
+
+    Method(
+      "release",
+      [this](dynamic args)  // blobId: string
+      {
+        auto blobId = jsArgAsString(args, 0);
+
+        m_contentHandler->RemoveMessage(std::move(blobId));
+      }
+    )
+  };
+}
+// clang-format on
+
+#pragma endregion CxxModule overrides
+
+#pragma region IWebSocketModuleContentHandler overrides
+
+void BlobWebSocketModuleContentHandler::Register(int64_t socketID) noexcept /*override*/
+{
+  lock_guard<mutex> lock{m_socketIDsMutex};
+  m_socketIDs.insert(socketID);
+}
+
+void BlobWebSocketModuleContentHandler::Unregister(int64_t socketID) noexcept /*override*/
+{
+  lock_guard<mutex> lock{m_socketIDsMutex};
+  if (m_socketIDs.find(socketID) != m_socketIDs.end())
+    m_socketIDs.erase(socketID);
+}
+
+bool BlobWebSocketModuleContentHandler::IsRegistered(int64_t socketID) noexcept /*override*/
+{
+  lock_guard<mutex> lock{m_socketIDsMutex};
+  return m_socketIDs.find(socketID) != m_socketIDs.end();
+}
+
+vector<uint8_t>
```

Comment:
this makes a lot of unnecessary copies of the array. There is one copy inside of this method, then at the call sites there will be another because of either CreateFromByteArray or because of `insert(...)`. Use something like a vector_view (like string_view is to string).

Reply:
Looking into it...

Reply:
There does not appear to be a container type available in C++17 that would work as a "vector view".

Changing this type would span to the `IWebSocketResource` interface, which would affect functionality outside of the BlobModule's scope.

Filed #6093 for follow-up.

Resolving.

Reply:
We don't have std::span until C++ 20, but there are some containers we could use here. Consider `winrt::array_view<uint8_t>` or `std::basic_string_view<uint8_t>`

Reply:
But I really agree with some of the worries. If binary blobs are large (they could definitely be multi-MB), we seem to be doing many copies in different places, along with encoding decoding. Seems like we could cause hangs.

Reply:
`basic_string_view` could be an option.
As I mentioned, this change could end up spanning all the way to the `arrayBuffer` message type and to the WS resource interface. Hence, the follow-up task (I can hold the backport until that #6093 is resolved).

Regarding `winrt::array_iew`, I'd rather not use non-STD in the WS interface, to keep it as implementation-agnostic as possible.